### PR TITLE
Rework and expand the Items section

### PR DESCRIPTION
**Pull request by IchHabeHunger54**
Reworks and expands the Items section. Specifically, it:
- Rewrites the items page.
- Adds information on food, and further documentation on Item.Properties.
- Explains the difference between items and itemstacks.
- Adds a new article about tools and armor, including explanations for the TierSortingRegistry and the ToolAction system.
- Adds an article about NBT.
- Adds an article about mob effects and potions.

------------------
Preview URL: https://pr-39.neoforged-docs-previews.pages.dev

### CODE REVIEW

File: docs/items/index.md [hunk 1]
```diff
@@ -1,72 +1,127 @@
...
-Creating an Item
-----------------
+## What Even Is an Item?
 
-### Basic Items
+Before we get further into creating items, it is important to understand what an item actually is, and what distinguishes it from, say, a [block][block]. Let's illustrate this using an example:
 
-Basic items that need no special functionality (think sticks or sugar) do not need custom classes. You can create an item by instantiating the `Item` class with an `Item$Properties` object. This `Item$Properties` object can be made via the constructor and customized by calling its methods. For instance:
+- In the world, you encounter a dirt block and want to mine it. This is a **block**, because it is placed in the world. (Actually, it is not a block, but a blockstate. See the [Blockstates article][blockstates] for more detailed information.)
+- Once you have mined the block, it is removed (= replaced with an air block) and the dirt drops. The dropped dirt is an item **entity**. This means that like other entities (pigs, zombies, arrows, etc.), it can inherently be moved by things like water pushing on it, or burned by fire and lava.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:10]: I guess this is technically true, but it might be better practice to leave a link to loot tables when we say the block 'drops'.

File: docs/items/index.md [hunk 2]
```diff
@@ -1,72 +1,127 @@
...
+- Once you pick up the dirt item entity, it becomes an **item stack** in your inventory. An item stack is, simply put, an instance of an item with some extra information, such as the stack size.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:11]: I'm not the biggest fan of this wording, but I can't think of anything better. This should probably link to the section below with a huge important/warning about the mutability of stacks. The mutability makes it very difficult to perform operations since you could be affecting something which shouldn't be touched.

File: docs/items/index.md [hunk 3]
```diff
@@ -1,72 +1,127 @@
...
+- Item stacks are backed by their corresponding **item** (which is what we're creating). Items hold information that is the same across all items (for example, every iron sword has a max durability of 250), while item stacks hold information that can be different between two similar items (for example, one iron sword has 100 uses left, while another iron sword has 200 uses left).
```
Comment (ChampionAsh5357) [inline docs/items/index.md:12]: It might be best to equate this to an item stack is backed by an item the same way a blockstate is backed by a block. This is not 100% true, but it could be used as a rough comparison.

File: docs/items/index.md [hunk 4]
```diff
@@ -1,72 +1,127 @@
...
+- `setNoRepair` - Disables anvil and crafting grid repairing for this item. Unused in vanilla.
+- `rarity` - Sets the rarity of this item. Currently, this simply changes the item's color. `Rarity` is an enum consisting of the four values `COMMON` (white, default), `UNCOMMON` (yellow), `RARE` (aqua) and `EPIC` (light purple). Be aware that mods may add more rarity types.
+- `requiredFeatures` - Sets the required feature flags for this item. This is mainly used for vanilla's feature locking system in minor versions. It is discouraged to use this, unless you're integrating with a system locked behind feature flags by vanilla.
+- `food` - Sets the [`FoodProperties`][food] of this item.
+
+For examples, or to look at the various values used by Minecraft, have a look at the `Items` class.
+
+### Food
+
+The `Item` class provides default functionality for food items, meaning you don't need a separate class for that. To make your item edible, all you need to do is set the `FoodProperties` on it through the `food` method in `Item.Properties`.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:33]: It might be useful to point out the methods to determine whether an item is considered food or how food changes the default behavior.

File: docs/items/index.md [hunk 5]
```diff
@@ -1,72 +1,127 @@
...
+- `saturationMod` - The saturation modifier used in calculating the [saturation value][hunger] restored when eating this food. The calculation is `min(2 * nutrition * saturationMod, playerNutrition)`, meaning that using `0.5` will make the effective saturation value the same as the nutrition value.
+- `meat` - Whether this item should be considered meat or not. Used e.g. for determining if healing dogs with this food is possible.
+- `alwaysEat` - Whether this item can always be eaten, even if the hunger bar is full. `false` by default, `true` for golden apples and other items that provide bonuses beyond just filling the hunger bar.
+- `fast` - Whether fast eating should be enabled for this food. `false` by default, `true` for dried kelp in vanilla.
+- `effect` - Adds a `MobEffectInstance` to apply when eating this item. The second parameter denotes the probability of the effect being applied; for example, Rotten Flesh has an 80% chance of applying the Hunger effect when eaten. This method comes in two variants, one that takes in a supplier (for your own effects) and one that directly takes a `MobEffectInstance` (for vanilla effects).
```
Comment (ChampionAsh5357) [inline docs/items/index.md:42]: Do we still allow registry replacement? If not, ignore this comment.

File: docs/items/index.md [hunk 6]
```diff
@@ -1,72 +1,127 @@
...
+- `build` - Once you've set everything you want to set, call `build` to get a `FoodProperties` object for further use.
+
+For examples, or to look at the various values used by Minecraft, have a look at the `Foods` class.
+
+### More Functionality
+
+Directly using `Item` only allows for very basic items. If you want to add functionality, for example right-click interactions, a custom class that extends `Item` is required. The `Item` class has many methods that can be overridden to do different things; see the classes `Item` and `IItemExtension` for more information. See also the [Interaction Pipeline][interactionpipeline] for where to begin for right-click behavior.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:49]: I would rather this be more detailed, but it can be left to a separate PR if you so desire.

File: docs/items/index.md [hunk 7]
```diff
@@ -1,72 +1,127 @@
...
+
+### Resources
+
+If you register your item and get your item (via `/give` or through a [creative tab][creativetabs]), you will find it to be missing a proper model and texture. This is because textures and models  are handled by Minecraft's resource system.
+
+To apply a simple texture to an item, you must add an item model JSON and a texture PNG. See the section on [resources] for more information.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:55]: Missing link to page.

File: docs/items/index.md [hunk 8]
```diff
@@ -1,72 +1,127 @@
...
+
+## `ItemStack`s
```
Comment (ChampionAsh5357) [inline docs/items/index.md:57]: This section is too sparse. It should have the same level of information as block states, probably even more due to the  many, many issues of mutability regarding this.

File: docs/items/index.md [hunk 9]
```diff
@@ -1,72 +1,127 @@
...
-  // Add to ingredients tab
-  if (event.getTabKey() == CreativeModeTabs.INGREDIENTS) {
-    event.accept(ITEM);
-    event.accept(BLOCK); // Takes in an ItemLike, assumes block has registered item
-  }
+public static void buildContents(BuildCreativeModeTabContentsEvent event) {
+    // Is this the tab we want to add to?
+    if (event.getTabKey() == CreativeModeTabs.INGREDIENTS) {
+        event.accept(MyItemsClass.MY_ITEM);
+        event.accept(MyBlocksClass.MY_BLOCK); // Accepts an ItemLike. This assumes that MY_BLOCK has a corresponding item.
```
Comment (ChampionAsh5357) [inline docs/items/index.md:88]: Probably better to put the comment above the line.

File: docs/items/index.md [hunk 10]
```diff
@@ -1,72 +1,127 @@
...
+    }
 }
 ```
 
-You can also enable or disable items being added through a `FeatureFlag` in the `FeatureFlagSet` or a boolean determining whether the player has permissions to see operator creative tabs.
+The event also provides some extra information, such as `getFlags()` to get the list of enabled feature flags, or `hasPermissions()` to check if the player has permissions to view the operator items tab.
+
+:::note
+This method is for adding your items to Minecraft's tabs, or to other mods' tabs. To add items to your own tabs, see below.
+:::
```
Comment (ChampionAsh5357) [inline docs/items/index.md:97]: This should be at the top of the section and a link to the relevant section. Maybe the headers should also be changed to reflect this.

File: docs/items/tools.md [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+When creating a `Tier`, it is assigned a block tag containing blocks that need this tool (or a better one) to be broken. For example, the `minecraft:needs_iron_tool` tag contains Diamond Ores (among others), and the `minecraft:needs_diamond_tool` tag contains blocks like Obsidian and Ancient Debris.
+
+You can reuse one of these tags for your tool if you're fine with that. For example, if we wanted our copper tools to just be more durable stone tools, we'd pass in `BlockTags.NEEDS_STONE_TOOL`.
+
+Alternatively, we can create our own tag, like so:
+
+```java
+public static final TagKey<Block> NEEDS_COPPER_TOOL = TagKey.create(BuiltInRegistries.BLOCK.key(), new ResourceLocation(MOD_ID, "needs_copper_tool"));
+```
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:81]: Probably a link to the tags section.

File: docs/items/tools.md [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    TierSortingRegistry.registerTier(
+            COPPER_TIER,
+            //The name to use for internal resolution. May use the Minecraft namespace if appropriate.
+            new ResourceLocation("minecraft", "copper"),
+            //A list of tiers that are considered lower than the type being added. For example, stone is lower than copper.
+            //We don't need to add wood and gold here because those are already lower than stone.
+            List.of(Tiers.STONE),
+            //A list of tiers that are considered higher than the type being added. For example, iron is higher than copper.
+            //We don't need to add diamond and netherite here because those are already higher than iron.
+            List.of(Tiers.IRON)
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:117]: I think you should also pass in a resource location here to show off how to add soft dependencies on modded tiers.

File: docs/items/tools.md [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+    );
+}
+```
+
+If you want to check if a tier is applicable for a block state, you can call `TierSortingRegistry#isCorrectTierForDrops`.
+
+## Custom Tool Types
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:124]: Multitools (pickaxe, axe, shovel, hoe) typically extend the `AxeItem` class because of enchantments for the axe being hardcoded to a class check. Not sure if this should be in here, but I figured it's useful info.

File: docs/items/tools.md [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+
+Similar to tools, armor uses a tier system (although a different one). What is called `Tier` for tools is called `ArmorMaterial` for armors. Like above, this example shows how to add copper armor; this can be adapted as needed. For the vanilla values, see the `ArmorMaterials` enum.
+
+```java
+// We place copper somewhere between chainmail and iron.
+public static final ArmorMaterial COPPER_ARMOR_MATERIAL = new ArmorMaterial() {
+    // The name of the armor material. Mainly determines where the armor texture is.
+    @Override
+    public String getName() {
+        return "copper";
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:153]: Should have modid prefix, otherwise texture search will be in `minecraft` namespace.

File: docs/items/tools.md [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    // Override for StringRepresentable. Should generally return the same as getName().
+    @Override
+    public String getSerializedName() {
+        return "copper";
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:159]: Should have modid prefix to guarantee uniqueness. Better to have this delegate to `#getName`.

File: docs/items/tools.md [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+    public int getDurabilityForType(ArmorItem.Type type) {
+        return switch (type) {
+            case HELMET -> 2;
+            case CHESTPLATE -> 4;
+            case LEGGINGS -> 6;
+            case BOOTS -> 2;
+        };
+    }
+
+    // Returns the toughness value of the armor.
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:188]: This doesn't really explain what this value is.

File: docs/items/tools.md [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+    // Only diamond and netherite have values greater than 0 here, so we just return 0.
+    @Override
+    public float getToughness() {
+        return 0;
+    }
+
+    // Returns the knockback resistance value of the armor.
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:195]: This doesn't really explain what this value is.

File: docs/items/tools.md [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+        ArmorItem.Type.HELMET,
+        // The item properties. We don't need to set the durability here because ArmorItem handles that for us.
+        new Item.Properties()
+));
+public static final Supplier<ArmorItem> COPPER_CHESTPLATE = ITEMS.register("copper_chestplate", () -> new ArmorItem(...));
+public static final Supplier<ArmorItem> COPPER_LEGGINGS = ITEMS.register("copper_leggings", () -> new ArmorItem(...));
+public static final Supplier<ArmorItem> COPPER_BOOTS = ITEMS.register("copper_boots", () -> new ArmorItem(...));
+```
+
+Besides the usual resources, armors also need a worn armor texture that will be rendered over the player model when the armor is equipped. This texture must be located at `src/main/resources/assets/<mod_id>/textures/models/armor/<material>_layer_1.png` for the helmet, chestplate and boots textures, and in the same directory at `<material>_layer_2.png` for the leggings.
```
Comment (ChampionAsh5357) [inline docs/items/tools.md:240]: Probably a mention to `#getArmorTexture` here in case for a custom location that doesn't follow the naming scheme.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Really good so far! Needs a bit more explanation with the many annoying complexities of stacks and materials. Also added a few nitpicky comments, but if you can't think of anything better, feel free to ignore.

File: docs/items/index.md (hunk 5 above)
Comment (IchHabeHunger54) [reply docs/items/index.md:44]: I actually don't know? But regardless, what exactly is this comment aiming at?

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Few comments and nitpicks.

**Comment by IchHabeHunger54** [comment]
This is it, everything I consider to be in the scope of the PR is added. Reviews please.

File: docs/datastorage/nbt.md [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+NBT is a format introduced in the earliest days of Minecraft, written by Notch himself. It is widely used throughout the Minecraft codebase for data storage.
+
+## Specification
+
+The NBT spec is similar to the JSON spec, with a few differences:
+
+- Distinct types for bytes, shorts, longs and floats exist, suffixed by `b`, `s`, `l` and `f`, respectively, similar to how they would be represented in Java code.
+  - Doubles may also be suffixed with `d`, but this is not required, similar to Java code. The optional `i` suffix available in Java for integers is not permitted.
+  - All of these suffixes may also be capital letters. So for example, `64b` is the same as `64B`, and `0.5F` is the same as `0.5f`.
```
Comment (ChampionAsh5357) [inline docs/datastorage/nbt.md:11]: Not case sensitive?

File: docs/datastorage/nbt.md [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+There is one obvious exception, though: `ListTag`s. Working with these is special because when getting a list tag through `CompoundTag#getList`, you must also specify the list type. So getting a list of strings, for example, would work like this:
+
+```java
+ListTag list = tag.getList("SomeListHere", Tag.TAG_STRING);
+```
+
+Similarly, when creating a `ListTag`, you must also specify the list type during creation:
+
+```java
+ListTag list = new ListTag(List.of(), Tag.TAG_STRING);
```
Comment (ChampionAsh5357) [inline docs/datastorage/nbt.md:77]: Weird example due to how `ListTag` works. The type is mutable, so an empty list would have a tag type of `0` rather than the type of the object.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Just two nitpick comments, but looks fine. Will ping for any additional review, but otherwise can be merged in Monday afterward.

File: docs/datastorage/nbt.md (hunk 19 above)
Comment (IchHabeHunger54) [reply docs/datastorage/nbt.md:11]: If the Minecraft wiki is anything to go by, no.

Comment (ChampionAsh5357) [reply docs/datastorage/nbt.md:11]: Sorry, I mean this as a way to word the sentence. Like, 'the suffixes are case insensitive.'

**Review by ChampionAsh5357 — APPROVED** [review]
